src/workerd/api/tests/pipe-streams-test.js

```diff
@@ -0,0 +1,827 @@
+    await rejects(writer.write(enc.encode('hello')), { message: 'boom' });
+  },
+};
+
+// Test pipeTo error in JS Readable aborts Internal Writable when preventAbort = false
+export const pipeToJsToInternalErroredSource = {
+  async test() {
+    const enc = new TextEncoder();
+    const rs = new ReadableStream({
+      async pull() {
+        throw new Error('boom');
+      },
+    });
+    const { readable, writable } = new IdentityTransformStream();
+    const pipe = rs.pipeTo(writable);
+
+    ok(writable.locked);
+
+    const reader = readable.getReader();
+
+    await rejects(reader.read(), { message: 'boom' });
+
+    ok(!writable.locked);
+
+    // Attempts to use the writable from here on will fail with the same error.
+    const writer = writable.getWriter();
+    await rejects(writer.write(enc.encode('hello')), { message: 'boom' });
+
+    await rejects(pipe, { message: 'boom' });
+  },
+};
+
+// Test pipeThrough from internal readable to JS writable
+export const pipeThroughInternalToJs = {
+  async test(_ctrl, env) {
+    const response = await env.SERVICE.fetch('http://test/stream');
+    const dec = new TextDecoder();
+
+    const output = [];
+    const ws = new WritableStream({
+      write(chunk) {
+        output.push(dec.decode(chunk));
+      },
+    });
+
+    const ts = new TransformStream();
+    await response.body.pipeThrough(ts).pipeTo(ws);
+
+    strictEqual(output.join(''), 'hello world '.repeat(100));
+  },
+};
+
+// Test pipeTo from internal readable to JS writable
+export const pipeToInternalToJs = {
+  async test(_ctrl, env) {
+    const response = await env.SERVICE.fetch('http://test/stream');
+    const dec = new TextDecoder();
+
+    const output = [];
+    const ws = new WritableStream({
+      write(chunk) {
+        output.push(dec.decode(chunk));
+      },
+    });
+
+    await response.body.pipeTo(ws);
+
+    strictEqual(output.join(''), 'hello world '.repeat(100));
+  },
+};
+
+// Test pipeThrough from JS readable to JS writable
+export const pipeThroughJsToJs = {
+  async test() {
+    const enc = new TextEncoder();
+    const dec = new TextDecoder();
+
+    const chunks = [enc.encode('hello'), enc.encode('there')];
+    const rs = new ReadableStream({
+      pull(c) {
+        const chunk = chunks.shift();
+        if (chunk) {
+          c.enqueue(chunk);
+        } else {
+          c.close();
+        }
+      },
+    });
+
+    const ts = new TransformStream({
+      transform(chunk, controller) {
+        controller.enqueue(dec.decode(chunk).toUpperCase());
+      },
+    });
+
+    const output = [];
+    const ws = new WritableStream({
+      write(chunk) {
+        output.push(chunk);
+      },
+    });
+
+    await rs.pipeThrough(ts).pipeTo(ws);
+
+    strictEqual(output.join(''), 'HELLOTHERE');
+  },
+};
+
+// Test pipeThrough error in JS Readable does not abort Internal Writable when preventAbort = true
+export const pipeThroughJsToInternalErroredSourcePreventAbort = {
```

Comment:
Same confusion regarding the meaning of "internal" as `pipeToJsToInternalErroredSource`.